### DISCUSSION TITLE
How do I use kanidm for authentication from the browser using session cookies?

### DISCUSSION THREAD

**Discussion by TimoKramer** [kanidm/kanidm#2467] [General]
I read through a large portion of your documentation and and some Discussion and other sources. One thing I don't quite understand is:
How does Kanidm fit into web authentication? Do I use the kanidm-API to authenticate from my backend and return a session-cookie or is this something Kanidm does for me in one form or another?
Thanks

**Comment by yaleman** [comment]
I'd suggest using Kanidm as an [OAuth2](https://kanidm.github.io/kanidm/master/integrations/oauth2.html) authentication source, an industry-standard way of integrating authentication with external services.

You can also use LDAPS if your application is terrible and old 😄

**Comment by Firstyear** [comment]
Kanidm on it's own "doesn't do much". Other applications have to "trust" kanidm to do auth for them. The way we handle this is what we call "integrations" in the book. https://kanidm.github.io/kanidm/stable/integrations/oauth2.html 

The common ones are:

* Oauth2/OIDC - this allows web-applications like nextcloud to work. Some other web applications can "trust" auth from a provider like apache2 with mod_auth_oidc infront
* Linux/Posix machines - When the unix resolver is configured on a machine it allows you to ssh in or login as you would a local user
* RADIUS - this is a "specialised" authentication provider for network devices, VPNs and wifi. This is what is needed for things like WPA2 enterprise
* LDAP - A lot of applications support LDAP as an identity and password verification service. 

Hope that helps :)

**Comment by TimoKramer** [comment]
Thanks @Firstyear @yaleman. Clearly it helps reading into oauth2. I always thought it is strictly connected with JWT and the likes. But obviously it is possible to use it with a regular web app that just gets redirected: https://datatracker.ietf.org/doc/html/rfc6749#section-4.1

**Reply by TimoKramer** [reply]
So just to leave it here for completeness... htmx or other regular web apps can use oauth2 with a server-side authentication flow that redirects the browser to kanidm for authentication and redirects back to the server for it to return a session cookie in case of success. Trying to implement that now. Will keep updates coming here.
